components/user-self-registration/org.wso2.carbon.identity.user.registration.mgt/src/main/java/org/wso2/carbon/identity/user/registration/mgt/utils/GraphBuilder.java

```diff
@@ -0,0 +1,291 @@
+            }
+            stepContentMap.put(step.getId(), step);
+            setFirstNodeIfNeeded(flowConfig, stepNode);
+        }
+
+        updateNodeMappings(nodeMap, nodeEdges);
+        flowConfig.setNodeConfigs(nodeMap);
+        flowConfig.setNodePageMappings(stepContentMap);
+        return flowConfig;
+    }
+
+    private static String handleEndNode(Map<String, NodeConfig> nodeMap) {
+
+        NodeConfig userOnboardingNode = createUserOnboardingNode();
+        nodeMap.put(userOnboardingNode.getId(), userOnboardingNode);
+        return userOnboardingNode.getId();
+    }
+
+    private static NodeConfig processRedirectionStep(StepDTO step, Map<String, NodeConfig> nodeMap,
+                                                     List<NodeEdge> nodeMappings, String endNodeId)
+            throws RegistrationFrameworkException {
+
+        ActionDTO action = step.getData().getAction();
+        if (action == null) {
+            throw handleClientException(ERROR_CODE_ACTION_DATA_NOT_FOUND, step.getId());
+        }
+        if (!EXECUTOR.equals(action.getType())) {
+            throw handleClientException(ERROR_CODE_INVALID_ACTION_TYPE, action.getType(), step.getId());
+        }
+
+        NodeConfig redirectionNode = createTaskExecutionNode(step.getId(), action.getExecutor());
+        String nextNodeId = COMPLETE.equals(action.getNextId()) ? endNodeId : action.getNextId();
+        nodeMap.put(redirectionNode.getId(), redirectionNode);
+        nodeMappings.add(new NodeEdge(redirectionNode.getId(), nextNodeId, null));
+        return redirectionNode;
+    }
+
+    private static NodeConfig processViewStep(StepDTO step, Map<String, NodeConfig> nodeMap, List<NodeEdge> nodeEdges,
+                                              String endNodeId) throws RegistrationFrameworkException {
+
+        List<NodeConfig> stepNodes = new ArrayList<>();
+        List<ComponentDTO> components = step.getData().getComponents();
+        if (components == null || components.isEmpty()) {
+            throw handleClientException(ERROR_CODE_COMPONENT_DATA_NOT_FOUND, step.getId());
+        }
+        for (ComponentDTO component : components) {
+            processComponent(component, stepNodes, step.getId());
+        }
+        return handleTempNodesInStep(stepNodes, step, nodeMap, nodeEdges, endNodeId);
+    }
+
+    private static void processComponent(ComponentDTO component, List<NodeConfig> stepNodes, String stepId)
+            throws RegistrationFrameworkException {
+
+        if (FORM.equals(component.getType())) {
+            LOG.debug("Processing form component: " + component.getId());
+            for (ComponentDTO subComponent : component.getComponents()) {
+                processComponent(subComponent, stepNodes, stepId);
+            }
+        } else if (BUTTON.equals(component.getType()) && component.getAction() != null) {
```

Comment:
```suggestion
        } else if (BUTTON.equals(component.getType())) {
```

Reply:
This null check is not needed as the action is being checked for `null` in `validateStepActions` method.